Swarm crashed with panic

`swarm version 1.2.3 (eaa53c7)`

I got this in the logs:
```
time="2016-12-29T09:35:05Z" level=error msg="HTTP error: Container 9681280172a9401f63cada000578e8627c00358e15986357df7533cafd0f1766 is paused, unpause the container before exec\n" status=500 
panic: runtime error: slice bounds out of range

goroutine 303036 [running]:
bufio.(*Writer).flush(0xc83dd9fb80, 0x0, 0x0)
        /usr/local/go/src/bufio/bufio.go:562 +0x2a7
bufio.(*Writer).Flush(0xc83dd9fb80, 0x0, 0x0)
        /usr/local/go/src/bufio/bufio.go:551 +0x2d
net/http.(*response).Flush(0xc82b7d16b0)
        /usr/local/go/src/net/http/server.go:1204 +0x45
github.com/docker/swarm/api.(*eventsHandler).Handle(0xc8202e7e60, 0xc83c8afb00, 0x0, 0x0)
        /go/src/github.com/docker/swarm/api/events.go:120 +0xebd
github.com/docker/swarm/cluster.(*EventHandlers).Handle(0xc820300ea0, 0xc83c8afb00)
        /go/src/github.com/docker/swarm/cluster/event.go:42 +0xff
github.com/docker/swarm/cluster/swarm.(*Cluster).Handle(0xc8202b38f0, 0xc83c8afb00, 0x0, 0x0)
        /go/src/github.com/docker/swarm/cluster/swarm/cluster.go:116 +0x3b
github.com/docker/swarm/cluster.(*Engine).handler(0xc820080dc0, 0xc839289ce0, 0x4, 0xc83b375e40, 0x40, 0xc8326479e0, 0x53, 0xc839289cf0, 0x9, 0xc839289d00, ...)
        /go/src/github.com/docker/swarm/cluster/engine.go:1198 +0x2d2
github.com/docker/swarm/cluster.(*Engine).(github.com/docker/swarm/cluster.handler)-fm(0xc839289ce0, 0x4, 0xc83b375e40, 0x40, 0xc8326479e0, 0x53, 0xc839289cf0, 0x9, 0xc839289d00, 0x4, ...)
        /go/src/github.com/docker/swarm/cluster/engine.go:211 +0x4e
github.com/docker/swarm/cluster.(*EventsMonitor).Start.func2(0x7f98c20b90c8, 0xc838896880, 0xc8388695a0, 0xc839118f60, 0xc838e69b00)
        /go/src/github.com/docker/swarm/cluster/event_monitor.go:70 +0x202
created by github.com/docker/swarm/cluster.(*EventsMonitor).Start
        /go/src/github.com/docker/swarm/cluster/event_monitor.go:76 +0x246

goroutine 1 [chan receive, 953 minutes]:
github.com/docker/swarm/api.(*Server).ListenAndServe(0xc8202e7dd0, 0x0, 0x0)
        /go/src/github.com/docker/swarm/api/server.go:118 +0x383
github.com/docker/swarm/cli.manage(0xc820120fc0)
        /go/src/github.com/docker/swarm/cli/manage.go:328 +0x18fd
github.com/docker/swarm/vendor/github.com/codegangsta/cli.Command.Run(0xf68158, 0x6, 0xf680f8, 0x1, 0x0, 0x0, 0x0, 0x1006a10, 0x17, 0x0, ...)
        /go/src/github.com/docker/swarm/vendor/github.com/codegangsta/cli/command.go:137 +0x107e
github.com/docker/swarm/vendor/github.com/codegangsta/cli.(*App).Run(0xc820120d80, 0xc82000a0f0, 0x5, 0x5, 0x0, 0x0)
        /go/src/github.com/docker/swarm/vendor/github.com/codegangsta/cli/app.go:176 +0xfff
github.com/docker/swarm/cli.Run()
        /go/src/github.com/docker/swarm/cli/cli.go:68 +0x567
main.main()
        /go/src/github.com/docker/swarm/main.go:13 +0x14
...
```
[swarm-crash-20161229.zip](https://github.com/docker/swarm/files/677147/swarm-crash-20161229.zip)

@skryzhny Thanks for reporting the crash! The crash is a data race from the following 2 goroutines. Can you send Daemon logs for 20 seconds leading to this crash? If you have debug level enabled or reproduce steps, they would be very helpful. This part of code has been changed recently. I want to make sure we fix the right problem. 

```
panic: runtime error: slice bounds out of range

goroutine 303036 [running]:

bufio.(*Writer).flush(0xc83dd9fb80, 0x0, 0x0)
    /usr/local/go/src/bufio/bufio.go:562 +0x2a7
bufio.(*Writer).Flush(0xc83dd9fb80, 0x0, 0x0)
    /usr/local/go/src/bufio/bufio.go:551 +0x2d
net/http.(*response).Flush(0xc82b7d16b0)
    /usr/local/go/src/net/http/server.go:1204 +0x45
github.com/docker/swarm/api.(*eventsHandler).Handle(0xc8202e7e60, 0xc83c8afb00, 0x0, 0x0)
    /go/src/github.com/docker/swarm/api/events.go:120 +0xebd
github.com/docker/swarm/cluster.(*EventHandlers).Handle(0xc820300ea0, 0xc83c8afb00)
    /go/src/github.com/docker/swarm/cluster/event.go:42 +0xff
github.com/docker/swarm/cluster/swarm.(*Cluster).Handle(0xc8202b38f0, 0xc83c8afb00, 0x0, 0x0)
    /go/src/github.com/docker/swarm/cluster/swarm/cluster.go:116 +0x3b
github.com/docker/swarm/cluster.(*Engine).handler(0xc820080dc0, 0xc839289ce0, 0x4, 0xc83b375e40, 0x40, 0xc8326479e0, 0x53, 0xc839289cf0, 0x9, 0xc839289d00, ...)
    /go/src/github.com/docker/swarm/cluster/engine.go:1198 +0x2d2
github.com/docker/swarm/cluster.(*Engine).(github.com/docker/swarm/cluster.handler)-fm(0xc839289ce0, 0x4, 0xc83b375e40, 0x40, 0xc8326479e0, 0x53, 0xc839289cf0, 0x9, 0xc839289d00, 0x4, ...)
    /go/src/github.com/docker/swarm/cluster/engine.go:211 +0x4e
github.com/docker/swarm/cluster.(*EventsMonitor).Start.func2(0x7f98c20b90c8, 0xc838896880, 0xc8388695a0, 0xc839118f60, 0xc838e69b00)
    /go/src/github.com/docker/swarm/cluster/event_monitor.go:70 +0x202
created by github.com/docker/swarm/cluster.(*EventsMonitor).Start
    /go/src/github.com/docker/swarm/cluster/event_monitor.go:76 +0x246

```

```
goroutine 310863 [runnable]:
net/http.(*persistConn).readLoop.func1(0xc800000008, 0x114db18)
    /usr/local/go/src/net/http/transport.go:935
net/http.(*bodyEOFSignal).Close(0xc83deef800, 0x0, 0x0)
    /usr/local/go/src/net/http/transport.go:1316 +0x196
github.com/docker/swarm/vendor/github.com/docker/engine-api/client/transport/cancellable.(*notifyingReader).Close(0xc83883eb80, 0x0, 0x0)
    /go/src/github.com/docker/swarm/vendor/github.com/docker/engine-api/client/transport/cancellable/cancellable.go:107 +0x49
bufio.(*Writer).flush(0xc83dd9fb80, 0x0, 0x0)
    /usr/local/go/src/bufio/bufio.go:562 +0x2a7
bufio.(*Writer).Flush(0xc83dd9fb80, 0x0, 0x0)
    /usr/local/go/src/bufio/bufio.go:551 +0x2d
net/http.(*response).Flush(0xc82b7d16b0)
    /usr/local/go/src/net/http/server.go:1204 +0x45
github.com/docker/swarm/api.(*eventsHandler).Handle(0xc8202e7e60, 0xc83b6dc780, 0x0, 0x0)
    /go/src/github.com/docker/swarm/api/events.go:120 +0xebd
github.com/docker/swarm/cluster.(*EventHandlers).Handle(0xc820300ea0, 0xc83b6dc780)
    /go/src/github.com/docker/swarm/cluster/event.go:42 +0xff
github.com/docker/swarm/cluster/swarm.(*Cluster).Handle(0xc8202b38f0, 0xc83b6dc780, 0x0, 0x0)
    /go/src/github.com/docker/swarm/cluster/swarm/cluster.go:116 +0x3b
github.com/docker/swarm/cluster.(*Engine).handler(0xc820080dc0, 0xc83954eb60, 0x4, 0xc83aadb3c0, 0x40, 0xc82efff980, 0x53, 0xc83954eb70, 0x9, 0xc83954eb80, ...)
    /go/src/github.com/docker/swarm/cluster/engine.go:1198 +0x2d2
github.com/docker/swarm/cluster.(*Engine).(github.com/docker/swarm/cluster.handler)-fm(0xc83954eb60, 0x4, 0xc83aadb3c0, 0x40, 0xc82efff980, 0x53, 0xc83954eb70, 0x9, 0xc83954eb80, 0x4, ...)
    /go/src/github.com/docker/swarm/cluster/engine.go:211 +0x4e
github.com/docker/swarm/cluster.(*EventsMonitor).Start.func2(0x7f98c20b90c8, 0xc83883eb80, 0xc838842020, 0xc83da1c240, 0xc83de259e0)
    /go/src/github.com/docker/swarm/cluster/event_monitor.go:70 +0x202
created by github.com/docker/swarm/cluster.(*EventsMonitor).Start
    /go/src/github.com/docker/swarm/cluster/event_monitor.go:76 +0x246
```

Unfortunately we didn't enabled debug loglevel at that time.
Here is records before crash:
```
time="2016-12-29T08:22:16Z" level=error msg="HTTP error: Container 786413a3672fd14a5e77680b59a5e124890730041f28702fa1d11ccca1a47d6b is paused, unpause the container before exec\n" status=500 
time="2016-12-29T08:22:16Z" level=error msg="HTTP error: Container 786413a3672fd14a5e77680b59a5e124890730041f28702fa1d11ccca1a47d6b is paused, unpause the container before exec\n" status=500 
time="2016-12-29T09:02:18Z" level=error msg="HTTP error: Error response from daemon: Error downloading remote context http://JUST_SOME_URL: Got HTTP status code >= 400: 401 Unauthorized" status=500 
time="2016-12-29T09:35:05Z" level=error msg="HTTP error: Container 9681280172a9401f63cada000578e8627c00358e15986357df7533cafd0f1766 is paused, unpause the container before exec\n" status=500 
panic: runtime error: slice bounds out of range

goroutine 303036 [running]:
```

--- a/cluster/engine.go
+++ b/cluster/engine.go
@@ -214,6 +214,7 @@ func (e *Engine) StartMonitorEvents() {
 
 	go func() {
 		if err := <-ec; err != nil {
+			log.WithFields(log.Fields{"name": e.Name, "id": e.ID}).WithError(err).Error("error monitoring events, will restart")
 			if !strings.Contains(err.Error(), "EOF") {
 				// failing node reconnect should use back-off strategy
 				<-e.refreshDelayer.Wait(e.getFailureCount())
@@ -890,11 +891,6 @@ func (e *Engine) refreshLoop() {
 			lastSpecUpdatedAt = time.Now()
 		}
 
-		if !healthy {
-			e.eventsMonitor.Stop()
-			e.StartMonitorEvents()
-		}
-
 		err = e.RefreshContainers(false)
 		if err == nil {
 			// Do not check error as older daemon doesn't support this call
@@ -1407,6 +1403,7 @@ func (e *Engine) NetworkDisconnect(container *Container, network *Network, force
 	return e.RefreshNetworks()
 }
 
+//IsConnectionError returns true when err is connection problem
 func IsConnectionError(err error) bool {
 	// dockerclient defines ErrConnectionRefused error. but if http client is from swarm, it's not using
 	// dockerclient. We need string matching for these cases. Remove the first character to deal with

--- a/cluster/event_monitor.go
+++ b/cluster/event_monitor.go
@@ -1,8 +1,6 @@
 package cluster
 
 import (
-	"io"
-
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/events"
 	"github.com/docker/swarm/swarmclient"
@@ -38,10 +36,6 @@ func (em *EventsMonitor) Start(ec chan error) {
 					return
 				}
 			case err := <-errStream:
-				if err == io.EOF {
-					ec <- nil
-					return
-				}
 				ec <- err
 				return
 			case <-em.stopChan:

--- a/cluster/event_monitor.go
+++ b/cluster/event_monitor.go
@@ -25,9 +25,11 @@ func NewEventsMonitor(cli swarmclient.SwarmAPIClient, handler func(msg events.Me
 // Start starts the EventsMonitor
 func (em *EventsMonitor) Start(ec chan error) {
 	em.stopChan = make(chan struct{})
-	responseStream, errStream := em.cli.Events(context.Background(), types.EventsOptions{})
+	ctx, cancel := context.WithCancel(context.Background())
+	responseStream, errStream := em.cli.Events(ctx, types.EventsOptions{})
 
 	go func() {
+		defer cancel()
 		for {
 			select {
 			case event := <-responseStream:

